### PR TITLE
[docs] details on mapping dynamic outputs

### LINKED ISSUE #6928: [Documentation Feedback] Please add example of passing extra arguments to the docs for dynamic mapping and collect
From this Slack thread: https://dagster.slack.com/archives/C01U954MEER/p1638730392418700

It would be helpful to have this example:

```py
@job
def test():
    mynumber = get_number()
    my_op_2().map(lambda x: another_op(a=x, b=mynumber))
```

### PR DESCRIPTION
**Pull request by alangenfeld**
resolves #6928

## Test Plan

read the docs

### CODE REVIEW

**Comment by alangenfeld** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6140**  (this PR)

This comment was auto-generated by [Graphite](https://graphite.dev?utm_source=gh).

**Review by alangenfeld — COMMENTED** [review]
This level of detail felt excessive for the concepts page so here felt like the right place to me.

Very open to feedback on wording and placement for this.

File: docs/sphinx/sections/api/apidocs/dynamic.rst [hunk 1]
```diff
@@ -8,3 +8,73 @@ These APIs provide the means for a simple kind of *dynamic orchestration* — wh
 .. autoclass:: DynamicOut
 
 .. autoclass:: DynamicOutput
+
+
+Mapping Details
+---------------
+
+To establish dependencies on a dynamic output in a composition function like :py:class:`job` or :py:class:`graph`, the ``map`` function is used on the object representing the dynamic output. `map` takes a ``Callable`` which receives a single argument. This callable is evaluated once, and the :py:class:`op` or :py:class:`graph` that are passed the input argument will establish dependencies. The return value is captured and wrapped in an object that allows for subsequent ``map`` or ``collect`` calls.
```
Comment (dpeng817) [inline docs/sphinx/sections/api/apidocs/dynamic.rst:16]: like :py:class:`job` or :py:class:`graph` -> like ones _decorated_ by ...

Comment (dpeng817) [inline docs/sphinx/sections/api/apidocs/dynamic.rst:16]: I also find this statement generally hard to follow without an accompanying code snippet. I think the below snippets somewhat alleviate this, but I think a snippet showing a simple mapping in action would be helpful.

File: docs/sphinx/sections/api/apidocs/dynamic.rst [hunk 2]
```diff
@@ -8,3 +8,73 @@ These APIs provide the means for a simple kind of *dynamic orchestration* — wh
...
+Multiple outputs are returned via a ``namedtuple``
+
+.. code-block:: python
+
+    @job
+    def multiple():
+        # can unpack on assignment (order based)
+        a_values, b_values = multiple_dynamic_values()
+        a_values.map(process)
+        do_stuff(b_values.collect())
```
Comment (dpeng817) [inline docs/sphinx/sections/api/apidocs/dynamic.rst:67]: maybe also show namedtuple access of outputs here. While it's shown in the next code snippet, I don't think it's entirely clear

**Review by dpeng817 — CHANGES_REQUESTED** [review]
Generally looks good, I think there just need to be a few more code snippets

**Review by yuhan — COMMENTED** [review]
I think this level of details is worth including in https://docs.dagster.io/concepts/ops-jobs-graphs/jobs-graphs because 
1. rst is hard to maintain, as we don't have tests set up for bad links.
2. code snippets aren't under tests which means they could get outdated and confuse readers.
3. although i can't think of a perfect concept page for this info to fit it, i think it's more of the concept info hierarchy problem, so we should maybe re-org the giant Jobs&Graphs a bit. i could imagine we start to expand details for other graph shapes like conditional branching and fan in. (and i like this direction a lot!)


we could expand the [Dynamic Mapping & Collect example](https://docs.dagster.io/concepts/ops-jobs-graphs/jobs-graphs#dynamic-mapping--collect) or move it to somewhere like Screen Shot 2022-01-21 at 4 16 11 PM

**Review by yuhan — DISMISSED** [review]
q management

**Review by alangenfeld — COMMENTED** [review]
good push back on the initial pass on this - I think this direction is much better

**Comment by yuhan** [comment]
content looks great. could we add tests for the code snippets, e.g. make sure the jobs can run properly at minimum?

**Review by yuhan — APPROVED** [review]
